### PR TITLE
[#147] Send file to extension side

### PR DESCRIPTION
**Pull request by GuChad369**
This is the core-side implementation for sending the manifest file. 

Right now, it's just a basic structure, and I will continue to build and test it. 

Once I achieve success, I will ask for a review. 

Creating this pull request makes it convenient to ask questions and have discussions.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -35,9 +35,9 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
 message(STATUS "pgmoneta ${VERSION_STRING}")
 
-set(generation TRUE)
-set(check TRUE)
-set(container FALSE)
+set(generation FALSE)
+set(check FALSE)
+set(container TRUE)
```
Comment (jesperpedersen) [inline CMakeLists.txt:40]: Why did you change these ?

File: src/include/extension.h [hunk 2]
```diff
@@ -39,7 +39,8 @@ extern "C" {
 #include <stdbool.h>
 #include <stdlib.h>
 
-#define MAX_QUERY_LENGTH 1024
+#define MAX_QUERY_LENGTH 2048
+#define PGMONETA_CHUNK_SIZE 1024
```
Comment (jesperpedersen) [inline src/include/extension.h:43]: `CHUNK_SIZE 8192` ?

File: src/libpgmoneta/message.c [hunk 3]
```diff
@@ -3138,3 +3140,31 @@ decode_base64(const char* base64_data, int* decoded_len)
...
    return decoded_data;
 }
+
+int
+pgmoneta_send_file(SSL* ssl, int socket, char* username, char* source_path, char* target_path)
+{
+   return 1;
+}
+
+// static char*
```
Comment (jesperpedersen) [inline src/libpgmoneta/message.c:3150]: Comment

File: src/main.c [hunk 4]
```diff
@@ -631,6 +631,49 @@ main(int argc, char** argv)
       goto error;
    }
 
+   // printf("Start...\n");
```
Comment (jesperpedersen) [inline src/main.c:634]: Comment

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (GuChad369) [reply CMakeLists.txt:40]: It's just a basic structure, and it's convenient to test locally to avoid manually generating the manifest.

**Comment by GuChad369** [comment]
After I test it, I will ask for a review.

**Comment by GuChad369** [comment]
I used `pgmoneta_send_file` to send the manifest file. I ran some tests, and the file was sent successfully.

File: src/libpgmoneta/message.c [hunk 5]
```diff
@@ -66,6 +66,8 @@ static unsigned char* decode_base64(const char* base64_data, int* decoded_len);
 static char** get_paths(const char* data, int* count);
 static void extract_file_name(const char* path, char* file_name, char* file_path);
 
+// static char* encode_base64(char* buffer, int bytes_read);
```
Comment (jesperpedersen) [inline src/libpgmoneta/message.c:69]: Remove

File: src/include/extension.h [hunk 6]
```diff
@@ -39,7 +39,8 @@ extern "C" {
 #include <stdbool.h>
 #include <stdlib.h>
 
-#define MAX_QUERY_LENGTH 1024
+#define MAX_QUERY_LENGTH 9216
```
Comment (jesperpedersen) [inline src/include/extension.h:42]: Just do `16384`

File: src/libpgmoneta/message.c [hunk 7]
```diff
@@ -3098,4 +3100,80 @@ decode_base64(const char* base64_data, int* decoded_len)
...
+
+      // Send the file content in chunks
+      while ((bytes_read = fread(buffer, 1, sizeof(buffer), file)) > 0)
+      {
+         pgmoneta_base64_encode(buffer, bytes_read, &encode_chunk, &encoded_size);
+
+         pgmoneta_ext_send_file_chunk(ssl, socket, target_path, encode_chunk, &qr);
+         if (!qr)
+         {
+            pgmoneta_log_warn("Sending file: Send file chunk failed");
```
Comment (jesperpedersen) [inline src/libpgmoneta/message.c:3138]: _error

File: src/libpgmoneta/message.c [hunk 8]
```diff
@@ -3098,4 +3100,80 @@ decode_base64(const char* base64_data, int* decoded_len)
...
+         }
+         pgmoneta_free_query_response(qr);
+         free(encode_chunk);
+         encode_chunk = NULL;
+         qr = NULL;
+      }
+   }
+   else if (qr != NULL && qr->tuples != NULL && qr->tuples->data != NULL && qr->tuples->data[0] != NULL && qr->tuples->data[0][0] == 'f')
+   {
+      pgmoneta_log_warn("Sending file: User %s is not SUPERUSER", username);
```
Comment (jesperpedersen) [inline src/libpgmoneta/message.c:3149]: _error

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
See, if you can get more information about `SUPERUSER` into the documentation.

https://www.postgresql.org/docs/current/role-attributes.html

**Comment by GuChad369** [comment]
We have previously discussed the privilege issue. It's not the safest option to use superuser role. However, for now, it's just a temporary choice. We've already set up the structure `pgmoneta_ext_privilege()`, and in the future, we can add new functions in the extension to handle the privilege aspect.

**Comment by jesperpedersen** [comment]
Yes, but we need it in the user level documentation

**Comment by GuChad369** [comment]
We already have a functions manual to guide users regarding the privileges in the extension project.

**Comment by jesperpedersen** [comment]
Yes, but `SUPERUSER` is dangerous so we need to explain that

**Comment by jesperpedersen** [comment]
Rebase, please

**Comment by jesperpedersen** [comment]
Rebased, and merged.

Thanks for your contribution !
